core/src/main/java/org/elasticsearch/discovery/zen/UnicastZenPing.java

```diff
@@ -518,6 +503,33 @@ public void run() {
         }
     }
 
+    private Tuple<List<DiscoveryNode>, HashSet<DiscoveryNode>> buildNodesToPing() {
+        HashSet<DiscoveryNode> nodesToPingSet = new HashSet<>();
+
+        nodesToPingSet.addAll(hostsProvider.buildDynamicNodes());
+
+        // add all possible master nodes that were active in the last known cluster configuration
+        for (ObjectCursor<DiscoveryNode> masterNode : contextProvider.nodes().getMasterNodes().values()) {
+            nodesToPingSet.add(masterNode.value);
+        }
+
+        final List<DiscoveryNode> resolvedDiscoveryNodes;
+        try {
+            resolvedDiscoveryNodes = resolveDiscoveryNodes(
```

Comment:
Hi @bleskes , if we add `resolveDiscoveryNodes`, we can't guarantee *configured hosts first* ping, Does it make sense?